---
title: "Patagona Pixy - For calculating average nucleotide diversity (pi)"
author: "Jessie Williamson"
date: "2/14/2018; last revised 2024-04-10"
output: html_document
---

From Williamson et al. 2024, giant hummingbirds, *PNAS*. 

This is a super quick script to calculate pi, nucleotide diversity, from whole Patagona genomes. Input read in here was generated with: 
- Run patagona_invar.slurm script in patagona > EFG_invar. This gets vcf ready for pixy. 
- Then run pixy_pata.slurm also in patagona > EFG_invar. This actually runs pixy on our file to calculate nucleotide diversity. 
- Output is a .txt file, which I transferred from CARC to local. I opened this as a .csv and re-saved file to be able to read in here. 

See this website for working with pixy data: https://pixy.readthedocs.io/en/latest/output.html#working-with-pixy-output-data

The example on their site is confusing, but in brief, to calculate mean nucleotide diversity per population you want to sum count_diff per population, then sum count_comparisons, then take sum of diffs/sum of comparisons. So in formula language: 

pi = (Sum of count_diffs)/(sum of count_comparisons)

If I want to get mean nucleotide diversity for northern and southern, in pixy know that it would be inflated because we have pop structure
To calculate: 
Can either take: mean of nucleotide diversity for argentina and chile and then for Ecuador and Peru (phrase this way) 
Mean northern and mean southern (note that this refers to mean of each northern and southern population)


```{R, echo=FALSE}
# I set some GLOBAL R chunk options here.
#   (to hide this message add "echo=FALSE" to the code chunk options)

knitr::opts_chunk$set(comment = NA, message = FALSE, warning = FALSE, width = 100)
knitr::opts_chunk$set(fig.align = "center", fig.height = 4, fig.width = 6)

#knitr::opts_chunk$set(cache = TRUE, autodep=TRUE)
knitr::opts_chunk$set(cache = TRUE, autodep=TRUE)
```

---

```{R}
setwd("/Users/Jessie/MSBbirds Dropbox/Jessie Williamson/Rdirectory/Patagona")
```

# Load packages
```{R}
library(reshape)
library(car)
library(GGally)
library(Hmisc)
library(gridExtra)
library(stats)
library(gplots)
library(ggplot2)
library(lsmeans)
library(effects)
library(plyr)
library(dplyr)
library(gridExtra)
library(lattice)
library(survival)
library(AICcmodavg)
library(MuMIn)
library(stats4) 
library(PMCMR) # for Kruskal-Wallis post-hocs
library(fmsb)
library(faraway)
library(reshape2)
library(popbio)
library(arm)
library(ggfortify) # PCA
library(adegenet)
library(pegas)
```

---

# Load in pixy data 
```{R}
#rm(list=ls(all=TRUE)) # clear workspace 

# Read in data (cleaned and filtered in Patagona_DataWrangling.Rmd)
pixy <- read.csv("Genomics/pixy/pixy_output_pi_2022-04-16.csv", na.strings = c("", "NA"), stringsAsFactors = TRUE)

pi.all <- read.table("Genomics/pixy/patagona_genomes_vcftoolsfilter_10kthin_100complete_vcftools.windowed.pi", header=T)
pi.v <- read.table("Genomics/pixy/patagona_genomes_piestimate_withpixyfile_vcftools.windowed.pi", header=T)


## TAJIMA'S D DATA
taj.all <- read.table("Genomics/pixy/patagona_genomes_vcftoolsfilter_10kthin_100complete_vcftools.Tajima.D",header=T)
# ^^ this is a 10kb window dataset, though it's not named this way
taj.50 <- read.table("Genomics/pixy/patagona_genomes_vcftoolsfilter_10kthin_100complete_vcftools_TajD50Kb.Tajima.D",header=T)
taj.100 <- read.table("Genomics/pixy/patagona_genomes_vcftoolsfilter_10kthin_100complete_vcftools_TajD100Kb.Tajima.D",header=T)

taj.N <- read.table("Genomics/pixy/patagona_genomes_vcftoolsfilter_10kthin_100complete_vcftools_TajD10Kb_Northern.Tajima.D",
                    header=T)
taj.S <- read.table("Genomics/pixy/patagona_genomes_vcftoolsfilter_10kthin_100complete_vcftools_TajD10Kb_Southern.Tajima.D",
                    header=T)
taj.ec <- read.table("Genomics/pixy/patagona_genomes_vcftoolsfilter_10kthin_100complete_vcftools_TajD10Kb_Ecuador.Tajima.D",
                     header=T)
taj.peru <- read.table("Genomics/pixy/patagona_genomes_vcftoolsfilter_10kthin_100complete_vcftools_TajD10Kb_Peru.Tajima.D",
                       header=T)
taj.arg <- read.table("Genomics/pixy/patagona_genomes_vcftoolsfilter_10kthin_100complete_vcftools_TajD10Kb_Argentina.Tajima.D",
                      header=T)
taj.chile <- read.table("Genomics/pixy/patagona_genomes_vcftoolsfilter_10kthin_100complete_vcftools_TajD10Kb_Chile.Tajima.D",
                        header=T)

# ABBA/BABA
dstat <- read.csv("ABBA-BABA_summary_2022-04-19.csv", na.strings = c("", "NA"), stringsAsFactors = TRUE)
```


Really good pop gen pi and Tajima's D command line and R tutorial: https://www.york.ac.uk/res/dasmahapatra/teaching/MBiol_sequence_analysis/workshop4_2019.html

Differences between VCFtools and pixy: 
https://github.com/ksamuk/pixy/issues/14

Sliding window scan of Tajima's D: 
https://eacooper400.github.io/gen8900/exercises/tajd.html


# Calculate pi for each population (and compare pixy output with vcftools output)
A little clunky this way, but it works. 
```{r}
# Quick tests of each dataset to explore data 
hist(pixy$avg_pi, br=20)
hist(pi.v$PI, br=20)

boxplot(pixy$avg_pi,ylab="diversity")
boxplot(pi.v$PI, ylab="diversity")

# Chromsome 1 comparison between pixy and vcftools
pi.chr <- subset(pixy, chromosome == "NC_044244.1")
plot(pi.chr$window_pos_1, pi.chr$window_pos_2, xlab="position", ylab="diversity")

pi.chr1 <- subset(pi.v, CHROM == "NC_044244.1")
plot(pi.chr1$BIN_START, pi.chr1$BIN_END, xlab="position", ylab="diversity")

mean(pi.chr1$PI)
mean(pi.chr$avg_pi, na.rm=TRUE)


# Chromsome 2 comparison between pixy and vcftools
pi.chr2.pixy <- subset(pixy, chromosome == "NC_044245.1")
plot(pi.chr2.pixy$window_pos_1, pi.chr2.pixy$window_pos_2, xlab="position", ylab="diversity")

pi.chr2.v <- subset(pi.v, CHROM == "NC_044245.1")
plot(pi.chr2.v$BIN_START, pi.chr2.v$BIN_END, xlab="position", ylab="diversity")

mean(pi.chr2.pixy$avg_pi, na.rm=TRUE)
mean(pi.chr2.v$PI)


#### Calculate pi for each population

# Pi for Argentina
argentina <- pixy[which(pixy$pop == "Argentina"), ]
sum(is.na(argentina$count_diffs)) # 648 NAs 
arg.nona <- na.omit(argentina)

arg.sum.count.diff <- sum(arg.nona$count_diffs); arg.sum.count.diff
arg.sum.comparisons <- sum(arg.nona$count_comparisons); arg.sum.comparisons
arg.pi <- arg.sum.count.diff/arg.sum.comparisons; arg.pi

# Pi for Chile
chile <- pixy[which(pixy$pop == "Chile"), ]
sum(is.na(chile$count_diffs)) 
chile.nona <- na.omit(chile)

chile.sum.count.diff <- sum(chile.nona$count_diffs); chile.sum.count.diff
chile.sum.comparisons <- sum(chile.nona$count_comparisons); chile.sum.comparisons
chile.pi <- chile.sum.count.diff/chile.sum.comparisons; chile.pi

# Pi for Ecuador
ecuador <- pixy[which(pixy$pop == "Ecuador"), ]
sum(is.na(ecuador$count_diffs)) 
ecuador.nona <- na.omit(ecuador)

ecuador.sum.count.diff <- sum(ecuador.nona$count_diffs); ecuador.sum.count.diff
ecuador.sum.comparisons <- sum(ecuador.nona$count_comparisons); ecuador.sum.comparisons
ecuador.pi <- ecuador.sum.count.diff/ecuador.sum.comparisons; ecuador.pi

# Pi for Peru
peru <- pixy[which(pixy$pop == "Peru"), ]
sum(is.na(peru$count_diffs)) 
peru.nona <- na.omit(peru)

peru.sum.count.diff <- sum(peru.nona$count_diffs); peru.sum.count.diff
peru.sum.comparisons <- sum(peru.nona$count_comparisons); peru.sum.comparisons
peru.pi <- peru.sum.count.diff/peru.sum.comparisons; peru.pi
```

Argentina pi = 0.002270478
Chile pi = 0.002181954
Ecuador pi = 0.00204324
Peru pi = 0.002279147


# Calculate nucleotide diversity of northern and southern
```{r}
# See notes above and from 4/13/22 EFG meeting about using this method/calculating this way
southern.pi <- mean(chile.pi, arg.pi); round(southern.pi,5)
northern.pi <- mean(ecuador.pi, peru.pi); northern.pi
```

Southern pi = 0.002181954
Northern pi = 0.00204324

# Round estimates for paper
```{r}
format(arg.pi, scientific = TRUE)
format(chile.pi, scientific = TRUE)
format(peru.pi, scientific = TRUE)
format(ecuador.pi, scientific = TRUE)

format(southern.pi, scientific = TRUE)
format(northern.pi, scientific = TRUE)
```

**pi, converted to sci notation:**
Argentina: 2.270478e-03 (2.27 x 10^3)
Chile: 2.181954e-03 (2.18 x 10^3)
Peru: 2.279147e-03 (2.28 x 10^3)
Ecuador: 2.04324e-03 (2.04 x 10^3)
Northern: 2.04324e-03 (2.04 x 10^3)
Southern: 2.181954e-03 (2.18 x 10^3)




# Tajima's D
```{r}
library(plotrix) #std.error function

# Tajima's D

# Northern: 0.1802649
mean(taj.N$TajimaD, na.rm=TRUE) # 0.1802649
std.error(taj.N$TajimaD) # 0.004240203

#testing if different from zero
t.test(taj.N$TajimaD) # t = 42.513, df = 50512, p-value < 2.2e-16
wilcox.test(taj.N$TajimaD) # V = 768889285, p-value < 2.2e-16

# Southern: 0.1668048
mean(taj.S$TajimaD, na.rm=TRUE) # 0.1668048
std.error(taj.S$TajimaD) # 0.003994512

#testing if different from zero
t.test(taj.S$TajimaD) # t = 41.758, df = 59079, p-value < 2.2e-16
wilcox.test(taj.S$TajimaD) # V = 1037470022, p-value < 2.2e-16


mean(taj.ec$TajimaD, na.rm=TRUE) # 0.1809617
mean(taj.peru$TajimaD, na.rm=TRUE) # 0.1488731
mean(taj.arg$TajimaD, na.rm=TRUE) # 0.1055365
mean(taj.chile$TajimaD, na.rm=TRUE) # 0.1401197

# Ec: 0.1809617
# Peru: 0.1488731
# arg: 0.1055365
# Chile: 0.1401197


# Make a hist of Tajima's D
hist(taj.N$TajimaD, br=20)

mean(taj.100$TajimaD, na.rm=TRUE)
```



# Heterozygosity plot (Fig. S6, panel D)
```{r}
library(ggbeeswarm)

# Read in data
het <- read.csv("Genomics/UCE/Patagona_TissueUCE_Heterozygosity.csv", na.strings = c("", "NA"), stringsAsFactors = TRUE)
allhet <- read.csv("Genomics/UCE/Patagona_TissueUCE_and_Genome_Heterozygosity.csv", na.strings = c("", "NA"), stringsAsFactors = TRUE)

str(het)
levels(het$lineage)
het$lineage <- factor(het$lineage, levels = c("Southern", "Northern", "Hybrid"))

str(allhet)
levels(allhet$lineage)
allhet$lineage <- factor(allhet$lineage, levels = c("Southern", "Northern", "Hybrid"))


library(ggsignif)
library(ggbeeswarm)

(het.quasirandom <- ggplot(subset(allhet, lineage %in% c("Northern", "Southern", "Hybrid")), aes(x=lineage, y=Heterozygosity, colour=lineage)) + # Add shape=Clade here to get shapes to work properly
  #     ggplot(subset(df, dose %in% c("D0.5", "D1"))
  geom_point(size=1.5, alpha=0.80) +
  #geom_beeswarm(size = 3, cex = 3) + 
  geom_quasirandom(bandwidth=.1, size=1.5, alpha=0.80) + 
  # geom_signif(comparisons = list(c("Southern", "Northern")), # List pairwise comparisons in the order you want sig to appear
  #                   y_position=c(0.2), # ypos=c(height of 1st annotation, height of 2nd annotation, 3rd, etc)
  #                   map_signif_level=TRUE, tip_length=0.005) +
  scale_color_manual(values = c("#8B795E", "#00C5CD", "#F708CA")) + # purple=southern, orange=northern, gray=unknown
  theme_classic() + 
  labs(x="", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="Heterozygosity") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(legend.position = "none") + 
 # annotate("text", x=4000, y=24.2, size=3, label = "italic(R) ^ 2==0.31", parse = TRUE) + # R^2 from full model
  #ggtitle("A") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=12), axis.text.x=element_text(size=12), axis.title=element_text(size=14))
 )
ggsave(het.quasirandom, filename="HetPlot_Quasirandom_ByLineage_2023-04-04.pdf", height=4.5, width=3.5, units="in")
```


---


# Print environment for reproducibility
```{r}
sessionInfo() # List of packages and versions in use 
```

###########

## END 
